pkg/template/jsonpath.go

```diff
@@ -0,0 +1,155 @@
+package template
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"reflect"
+	"regexp"
+
+	"k8s.io/client-go/third_party/forked/golang/template"
+	"k8s.io/client-go/util/jsonpath"
+)
+
+var (
+	tektonVar  = regexp.MustCompile(`^\$\((.*)\)$`)
+	jsonRegexp = regexp.MustCompile(`^\{\.?([^{}]+)\}$|^\.?([^{}]+)$`)
+)
+
+func ParseJSONPath(input interface{}, expr string) (string, error) {
+	j := jsonpath.New("").AllowMissingKeys(false)
+	buf := new(bytes.Buffer)
+
+	//First turn the expression into fully valid JSONPath
+	expr, err := TektonJSONPathExpression(expr)
+	if err != nil {
+		return "", err
+	}
+
+	if err := j.Parse(expr); err != nil {
+		return "", err
+	}
+
+	fullResults, err := j.FindResults(input)
+	if err != nil { //
+		return "", err
+	}
+
+	// Copy of j.PrintResults
+	for ix := range fullResults {
+		if err := printResults(buf, fullResults[ix]); err != nil {
+			return "", err
+		}
+	}
+
+	return buf.String(), nil
+}
+
+// PrintResults writes the results into writer
+// This is a slightly modified copy of the original
+// j.PrintResults from k8s.io/client-go/util/jsonpath/jsonpath.go
+// in that it uses calls `textValue()` for instead of `evalToText`
+// This is a workaround for kubernetes/kubernetes#16707
+func printResults(wr io.Writer, results []reflect.Value) error {
+	for i, r := range results {
+		text, err := textValue(r)
+		if err != nil {
+			return err
+		}
+		if i != len(results)-1 {
+			text = append(text, ' ')
+		}
+		if _, err = wr.Write(text); err != nil {
```

Comment:
Not sure if this is worth doing or not, but thoughts on doing things this way as opposed to passing the writer through?

Reply:
What's the advantage? My thinking was to keep the copied code from the library (i.e. the printResults and the evalToText function) as close to the original as possible, try to upstream the fix in  the original library itself, and then at some point in the future remove the copy pasta!

Reply:
SGTM